test/core/ChainManager.test.js

```diff
@@ -0,0 +1,131 @@
+import SocketService from '../../src/core/SocketService';
+import ChainManager from '../../src/core/ChainManager';
+import {
+  listAllChains,
+  listAllSnapshots,
+  getBlockNumber,
+  mineBlock
+} from '../../src/core/ChainRequest';
+
+jest.setTimeout(10000);
+let socket, service, id, chain, exists, response, block;
+
+const options = {
+  accounts: 3,
+  block_mine_time: 0,
+  clean_on_stop: false
+};
+
+beforeEach(async () => {
+  socket = new SocketService();
+  service = new ChainManager(socket);
+  await socket.init();
+  await service.init();
+  id = await service.createChain({ ...options });
+});
+
+afterEach(async () => {
+  await service.clean();
+});
+
+test('service will create chain instance', async () => {
+  chain = service.chain(id);
+  expect(chain.id).toEqual(id);
+  expect(chain.name).toEqual(`chain:${id}`);
+  expect(chain.active).toBe(true);
+  expect(chain.accounts.length).toBe(options.accounts + 1);
+});
+
+test('service will stop chain instance', async () => {
+  expect(service.chain(id).active).toBe(true);
+  await service.chain(id).stop();
+  expect(service.chain(id).active).toBe(false);
+});
+
+test('service will restart chain instance', async () => {
+  await service.chain(id).stop();
+  expect(service.chain(id).active).toBe(false);
+  await service.chain(id).start();
+  chain = service.chain(id);
+  expect(chain.active).toBe(true);
+  expect(chain.accounts.length).toBe(options.accounts + 1);
+});
+
+test('service will check existence of chain', async () => {
+  exists = await service.exists(id);
+  expect(exists).toBe(true);
+  id = new Array(20).join('1');
+  exists = await service.exists(id);
+  expect(exists).toBe(false);
+});
+
+test('service will delete chain instance when clean_on_stop is true', async () => {
+  id = await service.createChain({ ...options, clean_on_stop: true });
+  exists = await service.exists(id);
+  expect(exists).toBe(true);
+  await service.removeChain(id);
+  exists = await service.exists(id);
+  expect(exists).toBe(false);
+});
+
+test('service will delete chain instance when clean_on_stop is false', async () => {
```

Comment:
i'm not clear how this test is different from the one above. `clean_on_stop` is not set but everything else is the same?

Reply:
Answer to this is basically explained in this [function](https://github.com/makerdao/testchain-client/pull/17/files#diff-fdb8018a898e21f073cc03187b7cab41R97)

When `clean_on_stop` is `true`, it need only be stopped first and the server will delete it. In the case where it is `false` we must ensure that the chain is stopped first and then is hard deleted via a DELETE request to the api. 

This caused some confusion and impeded our progress in the initial stages of developing the client so I wanted to ensure that we can essentially clean the server of chains so testing can run a lot smoother. I suppose it would be better to have this test in the test file for ChainObject/Chain.js which could be a task for later when we start testing for that.